### DISCUSSION TITLE
Static analysis of SWIG generated code

### DISCUSSION THREAD

**Discussion by behrisch** [swig/swig#3065] [Q&A] [answered]
Hi everyone,
I have a question about running tools like SonarCloud on your SWIG generated code. Is this something which should be done because it helps finding bugs or should I rather avoid it because the SWIG code is quite advanced and will just confuse the analyzer. I am also asking in the light of #2865 which I would like to resolve but I am little bit hesitant just to ignore all errors in SWIG files. Thanks for your help!

**Comment by behrisch** [comment]
I had another one in #3076 but decided for now to ignore all these warnings. I would still be interested how others handle this topic, though. Thank you.

**Reply by ojwb** [reply, marked as answer]
I think it really depends on the tool in question and how useful the points it raises are.

I'm not sure if "SonarCloud" is the same as "SonarQube" or not - even if not these examples of proposed patches, many of which we couldn't apply, are illustrative:

https://github.com/swig/swig/pulls?q=is%3Apr+SonarQube+is%3Aclosed

This particular static analysis tool seems quite opinionated and for example wants us to use `nullptr` instead of `NULL`, `using` instead of `typedef`, `auto` instead of explicit types.  While those are all reasonable opinions if you're writing C++17 or newer code, SWIG still aims to generate code that compiles as C++98 (unless wrapping features from newer standards versions) so we simply can't do any of those unconditionally.  We also generate C as well as C++, so can't use C++-specific features except in C++-specific contexts.  We potentially could do some of the above conditional on the C++ standard version, but that would require macros or conditionals which makes it much more debatable as an improvement.  Excessive use of macros or preprocessor conditionals is arguably a C++ code smell too...

In general you are quite likely need to tweak the rules used by any static analysis tool for an existing codebase.  There's an additional wrinkle with SWIG that it generates code, and that can end up mixing SWIG code and code from user typemaps, but I don't think it's reasonable to expect SWIG to generate code that's clean with every static analysis tool even if we only consider the default rule sets.

Clean patches that fix actual problems are useful, whether identified by static analysis or not.  We really want a regression test for any fix though, so expect to need to create one starting from the problem that the static analyser claims to have identified.

Patches which are effectively suppressing static analysis warnings are interesting but we don't want to make the code harder to maintain.

We'd probably also consider patches which add pragmas, magic comments or similar to turn off unhelpful warnings for a particular tool (we do this for some compiler warnings that are hard to suppress in any other way), though I'd worry a bit we could end up with a lot of such pragmas/comments. Also unless the static analyser is FOSS we can't easily check if they're even still useful, and we also can't check if a code change introduces new warnings from it.
